### PR TITLE
Fixes #9393 - Allow users to create container without starting it

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

File: test/functionals/containers_steps_controller_test.rb [hunk 1]
```diff
@@ -12,6 +12,7 @@ class StepsControllerTest < ActionController::TestCase
         .returns(@container)
       put :update, { :wizard_state_id => state.id,
                      :id => :environment,
+                     :start_on_create => true,
```
Comment (shlomizadok) [inline test/functionals/containers_steps_controller_test.rb]: @eLobato - on this test I could either add the ```:start_n_create => true``` or change to:
```ruby
Service::Containers.any_instance.expects(:create_container!).with(equals(state))
.returns(@container)
```

File: app/controllers/containers/steps_controller.rb [hunk 2]
```diff
@@ -41,9 +46,13 @@ def set_form
           @state.send(:"#{step}") || @state.send(:"build_#{step}"))
     end
 
-    def create_container
+    def create_container(start = true)
       @state.send(:"create_#{step}", params[:"docker_container_wizard_states_#{step}"])
-      container = (service = Service::Containers.new).start_container!(@state)
+      container = if start == true
```
Comment (ohadlevy) [inline app/controllers/containers/steps_controller.rb]: nitpick: remote == true? or maybe use its start.is_a TrueClass ?

File: app/controllers/containers/steps_controller.rb [hunk 3]
```diff
@@ -41,9 +46,13 @@ def set_form
...
+                    (service = Service::Containers.new).start_container!(@state)
+                  else
+                    (service = Service::Containers.new).create_container!(@state)
```
Comment (ohadlevy) [inline app/controllers/containers/steps_controller.rb]: nitpick: extract the service outside of the if statement?

File: app/models/service/containers.rb [hunk 4]
```diff
@@ -6,28 +6,41 @@ def errors
...
-                                          :priority => environment_variable.priority
-          end
-        end
-        Taxonomy.enabled_taxonomies.each do |taxonomy|
-          container.send(:"#{taxonomy}=", wizard_state.preliminary.send(:"#{taxonomy}"))
-        end
-
-        fail ActiveRecord::Rollback unless pull_image(container) && start_container(container)
+        container = create_container_object(wizard_state)
+        container.save!
```
Comment (ohadlevy) [inline app/models/service/containers.rb]: save! might generate exception, did you test it on failure?

File: app/models/service/containers.rb [hunk 5]
```diff
@@ -6,28 +6,41 @@ def errors
...
+        run_container(container)
+        destroy_wizard_state(wizard_state)
+        container
+      end
+    end
 
+    def create_container!(wizard_state)
+      ActiveRecord::Base.transaction do
```
Comment (ohadlevy) [inline app/models/service/containers.rb]: what is the diff between AR::Base.transacation and Model.transacation blocks?

File: app/views/containers/steps/environment.html.erb [hunk 6]
```diff
@@ -16,7 +16,10 @@
         <%= link_to_add_fields(_("Add environment variable"), f, :environment_variables,
                                'foreman_docker/common_parameters/environment_variable') %>
       </div>
-   </div>
+    </div>
+    <div class="row">
+      <%= check_box_tag(:start_on_create) + _('Start on creation ?') %>
```
Comment (ohadlevy) [inline app/views/containers/steps/environment.html.erb]: maybe change text to Run? (with a desc that if its not selected its pull only?)

File: app/models/service/containers.rb (hunk 4 above)
Comment (shlomizadok) [inline app/models/service/containers.rb]: I'd like ```save!``` to generate an exception (http://apidock.com/rails/ActiveRecord/Persistence/save%21), I want to make sure all validations are passed when saving. Rails will generate the correct error for the user if validations failed.

File: app/models/service/containers.rb (hunk 5 above)
Comment (shlomizadok) [inline app/models/service/containers.rb]: ```Model.transaction``` deals with one (the model's) transaction in the block, ```AR::Base.transaction``` deals with all the transactions in the block. Here we have two transactions: ```container.save!``` and ```destroy_wizard_state(wizard_state)```

File: app/views/containers/steps/environment.html.erb (hunk 6 above)
Comment (shlomizadok) [inline app/views/containers/steps/environment.html.erb]: Run? or Run on creation?

**Comment by shlomizadok** [comment]
@eLobato tests are failing due to changes in encryptable in foreman core. (will check it tomorrow)

**Comment by shlomizadok** [comment]
@eLobato ping

File: app/controllers/containers/steps_controller.rb [hunk 7]
```diff
@@ -45,9 +45,14 @@ def set_form
           @state.send(:"#{step}") || @state.send(:"build_#{step}"))
     end
 
-    def create_container
+    def create_container(start = true)
       @state.send(:"create_#{step}", params[:"docker_container_wizard_states_#{step}"])
-      container = (service = Service::Containers.new).start_container!(@state)
+      service = Service::Containers.new
+      container = if start.is_a? TrueClass
```
Comment (daviddavis) [inline app/controllers/containers/steps_controller.rb]: Why `if start.is_a? TrueClass` instead of `if start`?

**Comment by daviddavis** [comment]
@shlomizadok. Code looks good. Just had one small comment. Rebase and I'll test it and merge.

Thanks.

File: app/controllers/containers/steps_controller.rb (hunk 7 above)
Comment (shlomizadok) [inline app/controllers/containers/steps_controller.rb]: start is not a model attribute and therefore doesn't have the ? hook
e.g
```ruby
2.0.0-p481 :006 > start = true
 => true 
2.0.0-p481 :007 > start?
NoMethodError: undefined method `start?' for main:Object
```

**Comment by shlomizadok** [comment]
Thanks for reviewing @daviddavis !
rebased  :)

File: app/controllers/containers/steps_controller.rb (hunk 7 above)
Comment (daviddavis) [inline app/controllers/containers/steps_controller.rb]: The ? is from my comment and is not part of the code:

```
container = if start
```

File: app/models/service/containers.rb [hunk 8]
```diff
@@ -75,8 +90,9 @@ def load_dns(state, r)
       end
     end
 
-    def full_messages
```
Comment (daviddavis) [inline app/models/service/containers.rb]: This method needs to stay. I think it's causing your test errors.

Comment (shlomizadok) [inline app/models/service/containers.rb]: :+1: Thanks!

File: app/controllers/containers/steps_controller.rb (hunk 7 above)
Comment (shlomizadok) [inline app/controllers/containers/steps_controller.rb]: @daviddavis, thought the same (https://github.com/theforeman/foreman-docker/pull/103#discussion_r27043575) :)

File: app/views/containers/steps/environment.html.erb [hunk 9]
```diff
@@ -29,7 +29,10 @@
         <%= link_to_add_fields(_("Add DNS"), f, :dns,
                                'foreman_docker/common_parameters/dns') %>                               
       </div>
-   </div>
+    </div>
+    <div class="row">
+      <%= check_box_tag(:start_on_create) + _('Run ?') + content_tag(:span, _('If selected, the container will start after it is created'), :class => "help-block").html_safe %>
```
Comment (daviddavis) [inline app/views/containers/steps/environment.html.erb]: I think `Run Container` might be better than `Run ?` here.

**Comment by daviddavis** [comment]
The run container checkbox and label looks odd compared to the other checkboxes on the right hand of the page.

screen shot 2015-04-29 at 11 50 16 am

**Comment by daviddavis** [comment]
Tested and works great. Fix the UI and I'll merge.

**Comment by shlomizadok** [comment]
@daviddavis - Changed the UI. Let me know if that's what you had in mind.

**Comment by daviddavis** [comment]
Merged as d5d9d39ee8a6a1c09c1c530b7d082c00cece0d38

**Pull request closed without merging**